worker/raft/raftbackstop/worker.go

```diff
@@ -0,0 +1,257 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package raftbackstop
+
+import (
+	"bytes"
+
+	"github.com/hashicorp/raft"
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	"github.com/juju/pubsub"
+	"gopkg.in/juju/names.v2"
+	"gopkg.in/juju/worker.v1"
+
+	"github.com/hashicorp/go-msgpack/codec"
+	"github.com/juju/juju/pubsub/apiserver"
+	"github.com/juju/juju/worker/catacomb"
+)
+
+// RaftNode captures the part of the *raft.Raft API needed by the
+// backstop worker.
+type RaftNode interface {
+	State() raft.RaftState
+	GetConfiguration() raft.ConfigurationFuture
+}
+
+// This worker monitors the state of the raft cluster it's passed, and
+// if it detects that there is only one remaining API server that is
+// non-voting, will force-append a configuration message to the raft
+// log to enable it to recover. This allows us to handle the situation
+// where we had two remaining machines (a leader and a non-voting
+// follower) but the leader was removed.
+
+// Config holds the values needed by the worker.
+type Config struct {
+	Raft     RaftNode
+	LogStore raft.LogStore
+	Hub      *pubsub.StructuredHub
+	Logger   loggo.Logger
+	Tag      names.Tag
+}
+
+// Validate validates the raft worker configuration.
+func (config Config) Validate() error {
```

Comment:
Need a nil check for Logger also or else it will explode in the worker loop

Reply:
loggo.Logger is a struct rather than an interface, so it can't be nil.

Reply:
I could define a logger interface - the pubsub and presence workers do that. I will do that.